search/utils.py

```diff
@@ -0,0 +1,366 @@
+
+    This is because the actual search result has nested types, which
+    would be cumbersome to access from arbitrary calling functions.
+    """
+    def __init__(self, search):
+        """Get raw search result from Elasticsearch."""
+        self._search = search
+
+    def count(self):
+        """Total records matching the query."""
+        return self._search.count()
+
+    def page_count(self):
+        """Total number of result pages."""
+        count = self._search.count() / PAGE_LENGTH
+        if self._search.count() % PAGE_LENGTH > 0:
+            count += 1
+        return int(count)
+
+    def get_page(self, page):
+        """Return paginated results."""
+        start = (page - 1) * PAGE_LENGTH
+        end = start + PAGE_LENGTH
+        return self._search[start:end].execute().hits
+
+    def all(self):
+        """Return all results in a generator."""
+        for i in range(self.page_count()):
+            for hit in self.get_page(i+1):
+                yield hit
+
+    def __getitem__(self, i):
+        """Return result by index."""
+        return self._search[i:i+1].execute().hits[0]
+
+
+def create_mapping():
+    """
+    Create the Elasticsearch mapping.
+
+    Notes on the "index" values:
+        analyzed: normal, will be split on whitespace for indexing
+            (partial matches can be made during queries)
+        not_analyzed: no split; matches must be exact
+        no: not indexed at all, but will be returned in search results
+    """
+
+    # Create the index if it doesn't exist.
+    if not CONN.indices.exists(INDEX_NAME):
+        CONN.indices.create(INDEX_NAME)
+    # Delete the mapping if an older version exists.
+    if CONN.indices.exists_type(index=INDEX_NAME, doc_type=DOC_TYPE):
+        CONN.indices.delete_mapping(index=INDEX_NAME, doc_type=DOC_TYPE)
+
+    mapping = Mapping(DOC_TYPE)
+
+    mapping.field("course", "string", index="not_analyzed")
+    mapping.field("description_path", "string", index="no")
+    mapping.field("description", "string", index="analyzed")
+    mapping.field("preview_url", "string", index="no")
+    mapping.field("repository", "string", index="not_analyzed")
+    mapping.field("resource_type", "string", index="analyzed")
+    mapping.field("content_xml", "string", index="no")
+    mapping.field("content_stripped", "string", index="analyzed")
+    mapping.field("run", "string", index="not_analyzed")
+    mapping.field("titlesort", "string", index="no")
+    mapping.field("title", "string", index="analyzed")
+
+    mapping.field("xa_avg_grade", "float")
+    mapping.field("xa_histogram_grade", "float")
+    mapping.field("xa_nr_attempts", "integer")
+    mapping.field("xa_nr_views", "integer")
+
+    mapping.save(INDEX_NAME)
+
+    # If we've done this, we always want to re-index all existing
+    # LearningResource instances. This function will probably only
+    # ever be called by migrations.
+    index_resources(LearningResource.objects.all())
```

Comment:
I don't believe that can ever be a problem. Because:

Any change (to `LearningResource`, probably) which would require the mapping to be updated would have the mapping data migration in the same branch.

There will never be a case where two arbitrary branches are in the same release and one of them breaks the mapping in the other. If two branches made incompatible changes, their schema migrations would have to be coordinated anyway, so the migration ordering/numbering would already be handled.